File: apollo_dump/scan_ic.py
"""
scan_ic.py — 扫描内存中 ItemCode 50125461 的所有位置
用法: python scan_ic.py
"""
import sys, os, time, subprocess
from datetime import datetime
sys.stdout.reconfigure(encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
LOG_FILE = os.path.join(SCRIPT_DIR, f'scan_ic_{datetime.now().strftime("%Y%m%d_%H%M%S")}.txt')
log_lines = []

def log(msg):
    line = f'[{time.strftime("%H:%M:%S")}] {msg}'
    print(line)
    log_lines.append(line)

def save_log():
    with open(LOG_FILE, 'w', encoding='utf-8') as f:
        f.write('\n'.join(log_lines))
    print(f'\n日志已保存: {LOG_FILE}')

def find_pid():
    try:
        r = subprocess.run(
            ['tasklist', '/fi', 'IMAGENAME eq FreeStyle.exe', '/fo', 'csv', '/nh'],
            capture_output=True, text=True, timeout=10
        )
        for line in r.stdout.strip().split('\n'):
            if 'FreeStyle.exe' in line:
                return int(line.split(',')[1].strip('"'))
    except Exception as e:
        log(f'查找进程失败: {e}')
    return None

def main():
    pid = find_pid()
    if not pid:
        log('FreeStyle.exe 未运行，请先启动游戏')
        return 1
    log(f'PID={pid}')

    import frida
    session = frida.attach(pid)
    log('Frida 已附加')

    js_path = os.path.join(SCRIPT_DIR, 'scan_ic.js')
    with open(js_path, 'r', encoding='utf-8') as f:
        js_code = f.read()

    def on_msg(msg, data):
        if msg['type'] == 'error':
            log(f'[JS错误] {msg.get("description","")} @ line {msg.get("lineNumber","?")}')
            return
        if msg['type'] != 'send':
            return

        p = msg['payload']
        t = p.get('t', '')

        if t == 'info':
            log(f'  {p["msg"]}')
        elif t == 'scan_result':
            log(f'  游戏模块内命中: {p["count"]} 处')
        elif t == 'hit':
            log(f'  [{p["n"]:2d}] {p["addr"]}  ({p["mod"]} +{p["modOffset"]})')
            log(f'       上下文: {p["ctx"]}')
        elif t == 'heap_hit':
            log(f'  [堆{p["n"]:2d}] {p["addr"]}  (size={p["size"]} prot={p["prot"]})')
            log(f'       上下文: {p["ctx"]}')
        elif t == 'done':
            log(f'  {p["msg"]}')
        elif t == 'heap_done':
            log(f'  堆内存命中: {p["count"]} 处')

    script = session.create_script(js_code)
    script.on('message', on_msg)
    script.load()

    log('扫描中，等待完成... (Ctrl+C 退出)\n')

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        pass

    script.unload()
    session.detach()
    log('已断开')
    save_log()
    return 0

if __name__ == '__main__':
    try:
        sys.exit(main())
    except Exception as e:
        log(f'异常: {e}')
        save_log()

File: apollo_dump/scan_ic.js
// scan_ic.js
// 扫描内存中所有包含 ItemCode 50125461 的位置

var SRC_IC = 50125461;
var DST_IC = 50125711;

// 50125461 = 0x02FCDA95 (小端: 95 DA FC 02)
var icBytes = [];
icBytes.push(SRC_IC & 0xFF);
icBytes.push((SRC_IC >> 8) & 0xFF);
icBytes.push((SRC_IC >> 16) & 0xFF);
icBytes.push((SRC_IC >> 24) & 0xFF);
var pattern = '';
for (var i = 0; i < icBytes.length; i++) {
    if (i > 0) pattern += ' ';
    pattern += ('0' + icBytes[i].toString(16)).slice(-2);
}

send({t:'info', msg:'扫描 ItemCode ' + SRC_IC + ' (0x' + SRC_IC.toString(16) + ')'});
send({t:'info', msg:'搜索模式: ' + pattern});

// 获取游戏模块范围
var gameMod = Process.getModuleByName('FreeStyle.exe');
send({t:'info', msg:'FreeStyle.exe base=' + gameMod.base + ' size=' + gameMod.size});

var allHits = [];

// 扫描游戏模块的 .text + .data
Memory.scan(gameMod.base, gameMod.size, pattern, {
    onMatch: function (address, size) {
        var offset = address.sub(gameMod.base);
        var mod = Process.findModuleByAddress(address);
        var modName = mod ? mod.name : '???';
        var modOffset = mod ? address.sub(mod.base) : '?';

        // 读取前后上下文（前后各8字节）
        var ctxBefore = new Uint8Array(address.sub(8).readByteArray(8));
        var ctxAfter = new Uint8Array(address.add(4).readByteArray(8));

        var hexBefore = '';
        for (var i = 0; i < ctxBefore.length; i++) hexBefore += ('0' + ctxBefore[i].toString(16)).slice(-2) + ' ';
        var hexAfter = '';
        for (var i = 0; i < ctxAfter.length; i++) hexAfter += ('0' + ctxAfter[i].toString(16)).slice(-2) + ' ';

        allHits.push({
            addr: address.toString(),
            offset: offset.toString(),
            mod: modName,
            modOffset: modOffset.toString(),
            before: hexBefore,
            after: hexAfter
        });
    },
    onComplete: function () {
        send({t:'scan_result', count: allHits.length});

        for (var i = 0; i < allHits.length; i++) {
            var h = allHits[i];
            send({t:'hit',
                n: i + 1,
                addr: h.addr,
                offset: h.offset,
                mod: h.mod,
                modOffset: h.modOffset,
                ctx: h.before + '| ' + pattern.replace(/ /g,' ') + ' |' + h.after
            });
        }

        send({t:'done', msg:'扫描完成，共 ' + allHits.length + ' 处'});

        // 额外: 也扫一下所有可读内存段（不在 FreeStyle.exe 模块内的）
        send({t:'info', msg:'继续扫描堆内存...'});
        var heapHits = 0;

        Process.enumerateRanges('r--').forEach(function (range) {
            // 跳过游戏模块（已扫过）和系统 DLL
            var rm = Process.findModuleByAddress(range.base);
            if (rm && rm.name === 'FreeStyle.exe') return;
            if (rm) return; // 跳过所有已加载模块

            try {
                Memory.scan(range.base, range.size, pattern, {
                    onMatch: function (address, size) {
                        heapHits++;
                        var ctxBefore = new Uint8Array(address.sub(8).readByteArray(8));
                        var ctxAfter = new Uint8Array(address.add(4).readByteArray(8));
                        var hexBefore = '';
                        for (var i = 0; i < ctxBefore.length; i++) hexBefore += ('0' + ctxBefore[i].toString(16)).slice(-2) + ' ';
                        var hexAfter = '';
                        for (var i = 0; i < ctxAfter.length; i++) hexAfter += ('0' + ctxAfter[i].toString(16)).slice(-2) + ' ';

                        send({t:'heap_hit',
                            n: heapHits,
                            addr: address.toString(),
                            size: range.size,
                            prot: range.protection,
                            ctx: hexBefore + '| ' + pattern.replace(/ /g,' ') + ' |' + hexAfter
                        });
                    },
                    onComplete: function () {}
                });
            } catch (e) {
                // 某些内存区域不可读，跳过
            }
        });

        send({t:'heap_done', count: heapHits});
    }
});
